futures-core/src/task_impl/mod.rs

```diff
@@ -15,69 +15,64 @@ pub use self::std::*;
 #[cfg(not(feature = "std"))]
 pub use self::core::*;
 
-pub struct BorrowedTask<'a> {
-    unpark: BorrowedUnpark<'a>,
-    // Task-local storage
+pub struct ContextInner<'a> {
+    unpark: TaskUnpark,
     map: &'a LocalMap,
 }
 
-fn with<F: FnOnce(&BorrowedTask) -> R, R>(f: F) -> R {
-    unsafe {
-        let task = get_ptr().expect("no Task is currently running");
-        assert!(!task.is_null(), "no Task is currently running");
-        f(&*(task as *const BorrowedTask))
-    }
-}
-
-/// A handle to a "task", which represents a single lightweight "thread" of
-/// execution driving a future to completion.
+/// The context for the currently running task.
+///
+/// A task represents a single lightweight "thread" of execution driving a
+/// future to completion.
 ///
 /// In general, futures are composed into large units of work, which are then
 /// spawned as tasks onto an *executor*. The executor is responsible for polling
 /// the future as notifications arrive, until the future terminates.
 ///
-/// This is obtained by the `task::current` function.
+/// The poll function accepts a context, which is usually provided by the executor.
+pub struct Context<'a>(Option<ContextInner<'a>>);
```

Comment:
This is certainly quite interesting! I think that for efficiency we will probably want the `'a` lifetime parameter on `Context`, but I could also see how that's not always desired (and it doesn't follow 100% for me in terms of what I might naively expect).

@cramertj do you have thoughts on this? Do you think we should avoid this entirely? Or in general, do you think that the concrete implementation of `Context` has an "obvious" implementation we should otherwise be doing?

Reply:
I had tossed around a few ideas here: I think the actual function signature we want is `mut ctx: Context<'a>`. `Context` does need to take the locally provided impls of `Spawn` etc. by reference, as well as the "super" `Context`, so we do need a lifetime parameter.

Reply:
Agreed!